src/history/hs.c

```diff
@@ -394,22 +394,54 @@ __hs_insert_updates_verbose(WT_SESSION_IMPL *session, WT_BTREE *btree)
  */
 static int
 __hs_insert_record_with_btree_int(WT_SESSION_IMPL *session, WT_CURSOR *cursor, WT_BTREE *btree,
-  const WT_ITEM *key, const WT_UPDATE *upd, const uint8_t type, const WT_ITEM *hs_value,
+  WT_ITEM *key, const WT_UPDATE *upd, const uint8_t type, const WT_ITEM *hs_value,
   WT_HS_TIME_POINT *stop_time_point)
 {
     WT_CURSOR_BTREE *cbt;
+    WT_DECL_ITEM(hs_key);
     WT_DECL_RET;
     WT_UPDATE *hs_upd, *upd_local;
+    wt_timestamp_t hs_start_ts;
+    uint64_t counter, hs_counter;
+    uint32_t hs_btree_id;
+    int cmp;
 
     cbt = (WT_CURSOR_BTREE *)cursor;
     hs_upd = upd_local = NULL;
 
+    /* Allocate buffers for the data store and history store key. */
+    WT_ERR(__wt_scr_alloc(session, 0, &hs_key));
+
+    /*
+     * Adjust counter if there exists an update in the history store with same btree id, key and
+     * timestamp. Otherwise the newly inserting history store record may fall behind the existing
+     * one can lead to wrong order.
+     */
+    counter = __wt_atomic_add64(&btree->hs_counter, 1);
+    WT_ERR_NOTFOUND_OK(
+      __wt_hs_cursor_position(session, cursor, btree->id, key, upd->start_ts), true);
+    if (ret == 0) {
+        WT_ERR(cursor->get_key(cursor, &hs_btree_id, hs_key, &hs_start_ts, &hs_counter));
+
+        /*
+         * Check the whether the existing record is also from the same timestamp.
+         *
+         * Verify simple checks first to confirm whether the retrieved update same or not before
+         * performing the expensive key comparison.
+         */
+        if (hs_btree_id == btree->id && upd->start_ts == hs_start_ts && counter < hs_counter) {
+            WT_ERR(__wt_compare(session, NULL, hs_key, key, &cmp));
+            if (cmp == 0)
+                counter = hs_counter + 1;
+        }
+        cursor->reset(cursor);
```

Comment:
The above approach fixes the problem. I thought of trying of the approach of traversing to the next record once the row_search is returned to check whether the next record from the same timestamp or not? But I am not sure whether that approach has any concurrency problems. 

If you find the above is fine, is there any way that we can reuse the search page in the later part?

Reply:
Why you need to reset the cursor?

Reply:
If we don't reset the cursor, we still hold the pin on the search page, even if we are doing another search and reset it later, still, it produces the leak warning during session close.

Reply:
I think there is a layering issue here because we're mixing cursor calls (`search_near` and `prev` in `__wt_hs_cursor_position`) with low-level operations like `__wt_row_search` below.  Only the cursor calls consider whether there is a page pinned that needs to be released.

To avoid a second search, we need logic like this: https://github.com/wiredtiger/wiredtiger/blob/f0f6731a2cca9aa7dddd872884b16438b9ec9c76/src/btree/bt_cursor.c#L637-L659

That is, we need to first attempt the search on the leaf page, and if that fails for some reason, release it and search again.  The best way to do that is probably to add some logic to `__wt_hs_search` that starts `if (cbt->ref != NULL)` or `if (F_ISSET(cbt, WT_CBT_ACTIVE))` ....

Reply:
Thanks. I changed the code to search the leaf page that was found in the earlier search to insert.